### PR TITLE
[Feature] Add TTSR.

### PR DESCRIPTION
**Pull request by Yshuo-Li**
Wait for ttsr_net, lte and search_transformer

### CODE REVIEW

File: mmedit/models/restorers/ttsr.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+            elif iteration is None:
+                save_path = osp.join(save_path, f'{folder_name}.png')
+            else:
+                raise ValueError('iteration should be number or None, '
+                                 f'but got {type(iteration)}')
+            mmcv.imwrite(tensor2img(pred), save_path)
+
+        return results
+
+    def val_step(self, data_batch, **kwargs):
```
Comment (ckkelvinchan) [inline mmedit/models/restorers/ttsr.py:226]: This can be inherited from `BasicRestorer`.

File: mmedit/models/restorers/ttsr.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        # model
+        self.generator = build_backbone(generator)
+        self.transformer = build_component(transformer)
+        self.extractor = build_component(extractor)
+
+        # loss
+        self.pixel_loss = build_loss(pixel_loss) if pixel_loss else None
+
+        # pretrained
+        if pretrained:
```
Comment (ckkelvinchan) [inline mmedit/models/restorers/ttsr.py:51]: This condition is checked in `init_weights`.

File: mmedit/models/restorers/ttsr.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        gt = data_batch['gt']
+        ref = data_batch['ref']
+        ref_downup = data_batch['ref_downup']
+
+        # generate
+        pred = self.forward_dummy(lq, lq_up, ref, ref_downup)
+
+        # loss
+        losses = dict()
+        log_vars = dict()
```
Comment (ckkelvinchan) [inline mmedit/models/restorers/ttsr.py:132]: This line is not necessary.

File: mmedit/models/restorers/ttsr.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+                Defaults to True.
+        """
+        if isinstance(pretrained, str):
+            if self.generator:
+                self.generator.init_weights(pretrained, strict)
+            if self.extractor:
+                self.extractor.init_weights(pretrained, strict)
+            if self.transformer:
+                self.transformer.init_weights(pretrained, strict)
+        elif pretrained is None:
```
Comment (ckkelvinchan) [inline mmedit/models/restorers/ttsr.py:255]: This can be similified as 
```
elif pretrained is not None:            
   raise TypeError(...)
```

File: mmedit/models/restorers/ttsr.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+            s (Tensor): Soft-Attention tensor with shape (n, 1, h, w).
+            t_level3 (Tensor): Transferred HR texture T in level3.
+                (n, 4c, h, w)
+            t_level2 (Tensor): Transferred HR texture T in level2.
+                (n, 2c, 2h, 2w)
+            t_level1 (Tensor): Transferred HR texture T in level1.
+                (n, c, 4h, 4w)
+        """
+
+        _, _, lq_up_level3 = self.extractor((lq_up.detach() + 1.) / 2.)
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:76]: It means `No feedback here`

[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:76]: It means `No feedback here`

File: mmedit/models/restorers/ttsr.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    def forward_dummy(self, lq, lq_up, ref, ref_downup, only_pred=True):
+        """Forward of networks.
+
+        Args:
+            lq (Tensor): LQ image
+            lq_up (Tensor): Upsampled LQ image
+            ref (Tensor): Reference image
+            ref_downup (Tensor): Image generated by sequentially applying
+                bicubic down-sampling and up-sampling on reference image
+            only_pred (bool): Only return pred or not. Default: True
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:63]: OK

File: mmedit/models/restorers/ttsr.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+        Returns:
+            pred (Tensor): Predicted super-resolution results (n, 3, 4h, 4w)
+            s (Tensor): Soft-Attention tensor with shape (n, 1, h, w).
+            t_level3 (Tensor): Transferred HR texture T in level3.
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:68]: T: Transferred HR texture

[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:68]: T: Transferred HR texture

File: mmedit/models/restorers/ttsr.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        Args:
+            lq (Tensor): LQ image
+            lq_up (Tensor): Upsampled LQ image
+            ref (Tensor): Reference image
+            ref_downup (Tensor): Image generated by sequentially applying
+                bicubic down-sampling and up-sampling on reference image
+            only_pred (bool): Only return pred or not. Default: True
+
+        Returns:
+            pred (Tensor): Predicted super-resolution results (n, 3, 4h, 4w)
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:66]: It is not necessary since `pred` has also been used in other models of MMEditing

File: mmedit/models/restorers/ttsr.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+        # model
+        self.generator = build_backbone(generator)
+        self.transformer = build_component(transformer)
+        self.extractor = build_component(extractor)
+
+        # loss
+        self.pixel_loss = build_loss(pixel_loss) if pixel_loss else None
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:48]: OK

File: mmedit/models/restorers/ttsr.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py:68]: Yes I know, but it seems difficult to infer the meaning from the name.

File: mmedit/models/restorers/ttsr.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:63]: Changed to `predicted results`

File: mmedit/models/restorers/ttsr.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py:76]: `lq_up` is upsampled by a bicubic/bilinear upsampling? If yes there would be no gradient flowing back even if `.detach` is not used.

File: mmedit/models/restorers/ttsr.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+            lq (Tensor): Input lq images.
+            gt (Tensor): Ground-truth image. Default: None.
+            test_mode (bool): Whether in test mode or not. Default: False.
+            kwargs (dict): Other arguments.
+        """
+
+        if test_mode:
+            return self.forward_test(lq, gt=gt, **kwargs)
+
+        return self.forward_dummy(lq, **kwargs)
```
[the remark this answers is not in the document]
Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py:105]: I forgot to delete this, ignore this.

File: mmedit/models/restorers/ttsr.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:68]: `T` is used in the paper

[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:68]: texture_level3 may be too long, T only used for `Transferred HR texture`.

[the remark this answers is not in the document]
Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py:68]: Hm I think it is okay then. The docstring also explained it. Is it possible to state in the docstring the number of the equation using this variable (if any)?

File: mmedit/models/restorers/ttsr.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:76]: Ok.

File: mmedit/models/restorers/ttsr.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        optimizer.step()
+
+        log_vars.pop('loss')  # remove the unnecessary 'loss'
+        outputs = dict(
+            log_vars=log_vars,
+            num_samples=len(gt.data),
+            results=dict(
+                lq=lq.cpu(), gt=gt.cpu(), ref=ref.cpu(), output=pred.cpu()))
+
+        pred = None
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:151]: Emmm... copy from EDSR

File: mmedit/models/restorers/ttsr.py (hunk 1 above)
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:226]: Sure, I will remove the function later.

File: mmedit/models/restorers/ttsr.py (hunk 3 above)
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:132]: Yes, It is.

File: mmedit/models/restorers/ttsr.py (hunk 4 above)
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:255]: OK

File: mmedit/models/restorers/ttsr.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+                (n, c, 4h, 4w)
+        """
+
+        _, _, lq_up_level3 = self.extractor((lq_up.detach() + 1.) / 2.)
+        _, _, ref_downup_level3 = self.extractor(
+            (ref_downup.detach() + 1.) / 2.)
+        ref_level1, ref_level2, ref_level3 = self.extractor(
+            (ref.detach() + 1.) / 2.)
+
+        s, t_level3, t_level2, t_level1 = self.transformer(
```
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:82]: Soft-attention, it is similar to T

File: mmedit/models/restorers/ttsr.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:76]: I will try it

File: mmedit/models/restorers/ttsr.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py:80]: Maybe add a remark if possible.

File: mmedit/models/restorers/ttsr.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py]: Not sure if it is necessary to do so.

File: mmedit/models/restorers/ttsr.py (hunk 3 above)
Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py]: It is updated in line 137.

Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py]: Saw in the updates.

File: mmedit/models/restorers/ttsr.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ckkelvinchan) [reply mmedit/models/restorers/ttsr.py]: Fixed in updated version.

File: mmedit/models/restorers/ttsr.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Yshuo-Li) [reply mmedit/models/restorers/ttsr.py:80]: It has a remark in `mmedit/models/transformers/search_transformer.py`

**Review by innerlee — APPROVED** [review]
